### ISSUE TITLE
Functional Component Verbosity

### ISSUE THREAD

**Issue by andylamax** [JetBrains/kotlin-wrappers#317]
The current verbosity in creating functional components in kotlin-react is uncanny.

I would have to create the function
```kotlin
external interface DumbProps {
   var name: String
}
```
```kotlin
val DumbComp = functionalComponent<DumbProps> {
    div { +"Hi, I am ${it.name}" }
}
```
Then, I would have to create an extension for RBuilder like so
```kotlin
fun RBuilder.DumbComp(props: DumbProps,handler:RHandler<DumbProps>) = child(DumbComp,props,handler)
```
This makes the usage of functional components and hooks so verbose.

Advice: RBuilder should add a member function to ease use of functional components and hooks like so

```kotlin
class RBuilder{
  . . .
  operator fun <P : RProps> FunctionalComponent<P>.invoke(
    props: P = jsObject(),
    handler: RBuilder.(P) -> Unit
  ) = child(this, props, handler)
  . . .
}
```
So that now, I can just go and call
```kotlin
DumbComp {
   attrs.name = "Andy" 
}
```

**Comment by andylamax** [comment]
#319

**Comment by Skolotsky** [comment]
As i commented there it's question about correct function interface. And more correct pattern is to have wrapper like this
```kotlin
fun RBuilder.DumbComp(name: String,handler:RHandler<DumbProps>) = child(DumbComp) { 
    attrs {
        this.name = name
    } 
    handler()
}
```
And you will not be able to call this component without property `name`

**Comment by andylamax** [comment]
If that's the case, is it safe to write `DumbComp` like this?

```kotlin
fun RBuilder.DumbComp(name: String) = child(functionalComponent<DumbProps>{props->
   val count by useState(0) // hooks and effects here?
},jsObject<DumbProps>){}
```

Coz atleast I'll have a declaration and a Wrapper at the same time

**Comment by Skolotsky** [comment]
That form is totally wrong. It will cause creating new type of functional component on each call of DumbComp.
You still need to create functional component and make separate wrapper for it.
```kotlin
val dumbComp = functionalComponent<DumbProps>{ props->
   val count by useState(0) // hooks and effects here
}

fun RBuilder.DumbComp(name: String) = child(dumbComp){ // so dumbComp will be the same each time
  attrs {
    this.name = name
  }
}
```

**Comment by andylamax** [comment]
That is what I am trying to find a way to avoid.

**Comment by Skolotsky** [comment]
For now I don't know better way. We need compiler plugin or something like that to avoid such verbosity and preserve type safety...

**Comment by bnorm** [comment]
A compiler plugin can definitely solve this problem. I've built an example of one which is specific to Kotlin IR: https://github.com/bnorm/kotlin-react-function. Can take a function like the following and automatically generates the RProps interface, the component property, and rewrites the original function.

```kotlin
@RFunction
fun RBuilder.Hello(name: String) {
  +"Hello, $name"
}
```

**Comment by Skolotsky** [comment]
@bnorm, great! I'll take a look on that after migrating to new IR backend!

**Comment by turansky** [comment]
Since `pre.215` function component creation supported by `RBuilder` (like it was supported for `RClass`):

```Kotlin
val DumbComp = functionalComponent<DumbProps> {
    div { +"Hi, I am ${it.name}" }
}

val ContainerComp = functionalComponent<ContainerProps> {
    DumbComp {
        attrs.name = "Andy"
    }
}
```

Issue can be closed
cc @Leonya

**Comment by turansky** [comment]
Fixed in #505
